File: graph/floyd-warshall.cc
// https://atcoder.jp/contests/abc012/tasks/abc012_4
// https://github.com/drken1215/book_algorithm_solution/blob/master/codes/chap14/code_14_5.cpp

#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

int main() {
  const int INF = 1000 * 299 + 1;
  int N, M;
  cin >> N >> M;
  vector<vector<int> > dist(N, vector<int>(N, INF));
  for (int n = 0; n < N; n++) {
    dist[n][n] = 0;
  }
  for (int m = 0; m < M; m++) {
    int a, b, t;
    cin >> a >> b >> t;
    dist[a - 1][b - 1] = t;
    dist[b - 1][a - 1] = t;
  }
  // Floyd-Warshall algorithm
  for (int k = 0; k < N; k++) {
    for (int i = 0; i < N; i++) {
      for (int j = 0; j < N; j++) {
        dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j]);
      }
    }
  }
  vector<int> temp(N);
  for (int i = 0; i < N; i++) {
    temp.at(i) = *max_element(dist.at(i).begin(), dist.at(i).end());
  }
  cout << *min_element(temp.begin(), temp.end()) << endl;
  return 0;
}

File: graph/floyd-warshall.py
# https://atcoder.jp/contests/abc143/tasks/abc143_e
# https://github.com/drken1215/book_algorithm_solution/blob/master/codes/chap14/code_14_5.cpp
# Python3=TLE, PyPy3=AC

INF = 10 ** 9 * 300 + 1

N, M, L = map(int, input().split())
dp = [[INF] * N for n in range(N)]
for n in range(N):
    dp[n][n] = 0

for m in range(M):
    A, B, C = map(int, input().split())
    dp[A - 1][B - 1] = C
    dp[B - 1][A - 1] = C

# Floyd-Warshall algorithm
for k in range(N):
    for i in range(N):
        for j in range(N):
            dp[i][j] = min(dp[i][j], dp[i][k] + dp[k][j])
exist_negative_cycle = False
for n in range(N):
    if dp[n][n] < 0:
        exist_negative_cycle = True

dp2 = [[INF] * N for n in range(N)]
for i in range(N):
    for j in range(N):
        if i == j:
            dp2[i][j] = 0
        elif dp[i][j] <= L:
            dp2[i][j] = 1

# Floyd-Warshall algorithm
for k in range(N):
    for i in range(N):
        for j in range(N):
            dp2[i][j] = min(dp2[i][j], dp2[i][k] + dp2[k][j])

Q = int(input())
for q in range(Q):
    s, t = map(int, input().split())
    print(-1 if dp2[s - 1][t - 1] == INF else (dp2[s - 1][t - 1] - 1))
